File: python/src/heaton_life/ca/cyclic.py
"""Cyclic cellular automata (Griffeath) — spec/cyclic.md.

Each cell advances to the successor state (mod n) when at least `threshold`
neighbors within `reach` already hold that successor. Toroidal by definition.
"""

from __future__ import annotations

import dataclasses

import numpy as np
from numpy.typing import NDArray

from heaton_life.core.params import Params
from heaton_life.core.rng import Pcg32


@dataclasses.dataclass(frozen=True)
class CyclicParams(Params):
    states: int = dataclasses.field(default=14, metadata={"min": 2, "max": 24})
    threshold: int = dataclasses.field(default=1, metadata={"min": 1, "max": 48})
    reach: int = dataclasses.field(default=1, metadata={"min": 1, "max": 3, "label": "Range"})
    neighborhood: str = dataclasses.field(
        default="moore", metadata={"choices": ["moore", "vonneumann"]}
    )
    width: int = dataclasses.field(default=256, metadata={"min": 8, "max": 2048})
    height: int = dataclasses.field(default=256, metadata={"min": 8, "max": 2048})
    init: str = dataclasses.field(default="soup", metadata={"choices": ["soup"]})
    seed: int = dataclasses.field(
        default=0, metadata={"min": 0, "max": 4294967295, "role": "seed"}
    )


def _offsets(reach: int, neighborhood: str) -> list[tuple[int, int]]:
    out = []
    for dy in range(-reach, reach + 1):
        for dx in range(-reach, reach + 1):
            if (dy, dx) == (0, 0):
                continue
            if neighborhood == "vonneumann" and abs(dy) + abs(dx) > reach:
                continue
            out.append((dy, dx))
    return out


class Cyclic:
    """A cyclic CA. State is uint8 in [0, states), shape (height, width), torus."""

    def __init__(
        self,
        states: int = 14,
        *,
        size: tuple[int, int] = (256, 256),
        threshold: int = 1,
        reach: int = 1,
        neighborhood: str = "moore",
        init: str | NDArray[np.uint8] = "soup",
        seed: int = 0,
    ) -> None:
        if neighborhood not in ("moore", "vonneumann"):
            raise ValueError(f"unknown neighborhood: {neighborhood!r}")
        width, height = size
        self._initial: NDArray[np.uint8] | None = None
        if isinstance(init, np.ndarray):
            if init.shape != (height, width):
                raise ValueError(
                    f"init array shape {init.shape} does not match (h={height}, w={width})"
                )
            if init.max(initial=0) >= states:
                raise ValueError("init array has values >= states")
            self._initial = init.astype(np.uint8)
            init_name = "array"
        else:
            init_name = init
        self.params = CyclicParams(
            states=states,
            threshold=threshold,
            reach=reach,
            neighborhood=neighborhood,
            width=width,
            height=height,
            init=init_name,
            seed=seed,
        )
        self._generation = 0
        self._state: NDArray[np.uint8]
        self.reset()

    @classmethod
    def from_params(cls, params: CyclicParams) -> Cyclic:
        if params.init == "array":
            raise ValueError("params with init='array' need the array: Cyclic(init=...)")
        return cls(
            params.states,
            size=(params.width, params.height),
            threshold=params.threshold,
            reach=params.reach,
            neighborhood=params.neighborhood,
            init=params.init,
            seed=params.seed,
        )

    def reset(self, seed: int | None = None) -> None:
        if seed is not None:
            self.params = self.params.replace(seed=seed)
        p = self.params
        if p.init == "array":
            assert self._initial is not None
            self._state = self._initial.copy()
        elif p.init == "soup":
            draws = Pcg32(p.seed).fill_u32(p.width * p.height).reshape(p.height, p.width)
            self._state = (draws % np.uint32(p.states)).astype(np.uint8)
        else:
            raise ValueError(f"unknown init strategy: {p.init!r}")
        self._generation = 0

    def step(self, n: int = 1) -> None:
        p = self.params
        offsets = _offsets(p.reach, p.neighborhood)
        for _ in range(n):
            state = self._state
            succ = ((state.astype(np.uint16) + 1) % p.states).astype(np.uint8)
            count = np.zeros(state.shape, dtype=np.uint8)
            for dy, dx in offsets:
                count += np.roll(state, (dy, dx), axis=(0, 1)) == succ
            self._state = np.where(count >= p.threshold, succ, state)
        self._generation += n

    @property
    def state(self) -> NDArray[np.uint8]:
        return self._state

    @property
    def generation(self) -> int:
        return self._generation

    def frame(self) -> NDArray[np.uint8]:
        span = max(self.params.states - 1, 1)
        return (self._state.astype(np.uint16) * 255 // span).astype(np.uint8)

File: dotnet/src/HeatonLife.Core/Cyclic.cs
using System;
using System.Collections.Generic;

namespace HeatonLife
{
    public enum Neighborhood
    {
        Moore,
        VonNeumann,
    }

    /// <summary>
    /// Cyclic cellular automaton, Griffeath (spec/cyclic.md). Each cell advances to the
    /// successor state (mod States) when at least Threshold neighbors within Reach
    /// already hold that successor. State is a flat row-major byte grid in [0, States),
    /// toroidal by definition. Bit-exact tier.
    /// </summary>
    public sealed class Cyclic : IIndexedFrameSource
    {
        private readonly (int Dy, int Dx)[] _offsets;
        private byte[] _cells;
        private byte[] _scratch;

        public int States { get; }
        public int Threshold { get; }
        public int Reach { get; }
        public Neighborhood Neighborhood { get; }
        public int Width { get; }
        public int Height { get; }
        public int Generation { get; private set; }

        public ReadOnlySpan<byte> State => _cells;

        public Cyclic(
            int states,
            int width,
            int height,
            int threshold = 1,
            int reach = 1,
            Neighborhood neighborhood = Neighborhood.Moore)
        {
            if (states < 2)
                throw new ArgumentOutOfRangeException(nameof(states), $"states must be >= 2, got {states}");
            if (reach < 1)
                throw new ArgumentOutOfRangeException(nameof(reach), $"reach must be >= 1, got {reach}");
            States = states;
            Threshold = threshold;
            Reach = reach;
            Neighborhood = neighborhood;
            Width = width;
            Height = height;
            _offsets = BuildOffsets(reach, neighborhood);
            _cells = new byte[width * height];
            _scratch = new byte[width * height];
            SeedSoup(0); // the Python constructor's default init
        }

        private static (int, int)[] BuildOffsets(int reach, Neighborhood neighborhood)
        {
            var offsets = new List<(int, int)>();
            for (int dy = -reach; dy <= reach; dy++)
                for (int dx = -reach; dx <= reach; dx++)
                {
                    if (dy == 0 && dx == 0)
                        continue;
                    if (neighborhood == Neighborhood.VonNeumann && Math.Abs(dy) + Math.Abs(dx) > reach)
                        continue;
                    offsets.Add((dy, dx));
                }
            return offsets.ToArray();
        }

        /// <summary>Soup init per spec: one PCG32 draw per cell in row-major order, state = draw % States.</summary>
        public void SeedSoup(uint seed)
        {
            var rng = new Pcg32(seed);
            for (int i = 0; i < _cells.Length; i++)
                _cells[i] = (byte)(rng.NextU32() % (uint)States);
            Generation = 0;
            Remember(seed, s => SeedSoup(s));
        }

        // --- reset ---------------------------------------------------------------
        // How this world was last seeded, so Reset can replay it. The Python
        // reference keeps init/density/seed on its params object and re-dispatches
        // in reset(); a captured delegate is the same idea without a params type.
        private Action<uint> _replayInit = _ => { };
        private uint _seed;

        private void Remember(uint seed, Action<uint> replay)
        {
            _seed = seed;
            _replayInit = replay;
        }

        /// <inheritdoc />
        public void Reset(uint? seed = null)
        {
            if (seed.HasValue)
                _seed = seed.Value;
            _replayInit(_seed);
        }

        /// <summary>Load an explicit grid (row-major, Height*Width entries, values in [0, States)).</summary>
        public void SetState(ReadOnlySpan<byte> cells)
        {
            if (cells.Length != _cells.Length)
                throw new ArgumentException($"expected {_cells.Length} cells, got {cells.Length}");
            for (int i = 0; i < cells.Length; i++)
            {
                if (cells[i] >= States)
                    throw new ArgumentException($"cell value {cells[i]} >= states ({States})");
                _cells[i] = cells[i];
            }
            Generation = 0;
            var snapshot = _cells.Clone() as byte[];
            Remember(_seed, _ =>
            {
                Array.Copy(snapshot!, _cells, _cells.Length);
                Generation = 0;
            });
        }

        /// <summary>Restore a saved state at a given generation (catalog loads).</summary>
        public void SetState(ReadOnlySpan<byte> cells, int generation)
        {
            SetState(cells);
            Generation = generation;
        }

        /// <summary>Paint one cell in place (value in [0, States)); does not reset the generation.</summary>
        public void SetCell(int x, int y, byte value)
        {
            if (x < 0 || x >= Width || y < 0 || y >= Height)
                throw new ArgumentOutOfRangeException($"cell ({x}, {y}) outside {Width}x{Height}");
            if (value >= States)
                throw new ArgumentException($"cell value {value} >= states ({States})");
            _cells[y * Width + x] = value;
        }

        /// <summary>Restore a rectangle (inclusive corners) to the family blank, 0 (spec/patterns.md "Clear").</summary>
        public void ClearRect(int x0, int y0, int x1, int y1)
        {
            for (int y = y0; y <= y1; y++)
                for (int x = x0; x <= x1; x++)
                    SetCell(x, y, 0);
        }

        public void Step(int n = 1)
        {
            for (int s = 0; s < n; s++)
                StepOnce();
            Generation += n;
        }

        /// <summary>Frame per spec/render.md: palette index = state * 255 / max(States - 1, 1).</summary>
        public void WriteFrame(byte[] frame)
        {
            if (frame.Length != _cells.Length)
                throw new ArgumentException($"expected {_cells.Length} bytes, got {frame.Length}");
            int span = Math.Max(States - 1, 1);
            for (int i = 0; i < _cells.Length; i++)
                frame[i] = (byte)(_cells[i] * 255 / span);
        }

        /// <summary>Allocating convenience for <see cref="WriteFrame"/>.</summary>
        public byte[] Frame()
        {
            var frame = new byte[_cells.Length];
            WriteFrame(frame);
            return frame;
        }

        private void StepOnce()
        {
            int w = Width, h = Height;
            for (int y = 0; y < h; y++)
            {
                for (int x = 0; x < w; x++)
                {
                    byte current = _cells[y * w + x];
                    byte succ = (byte)((current + 1) % States);
                    int count = 0;
                    foreach (var (dy, dx) in _offsets)
                    {
                        int ny = (y + dy + h) % h;
                        int nx = (x + dx + w) % w;
                        if (_cells[ny * w + nx] == succ)
                            count++;
                    }
                    _scratch[y * w + x] = count >= Threshold ? succ : current;
                }
            }
            (_cells, _scratch) = (_scratch, _cells);
        }
    }
}
